includes/class-newspack-popups-inserter.php

```diff
@@ -293,26 +296,63 @@ public static function register_amp_scripts() {
 		}
 	}
 
+	/**
+	 * Some popups can only appear on Posts.
+	 *
+	 * @param object $popup The popup to assess.
+	 * @return bool Should popup be shown.
+	 */
+	public static function assess_is_post( $popup ) {
+		if (
+			// Inline Pop-ups can only appear in Posts.
+			'inline' === $popup['options']['placement'] ||
+			// Pop-ups triggered by scroll position can only appear on Posts.
+			'scroll' === $popup['options']['trigger_type']
+		) {
+			return is_single();
+		}
+		return true;
+	}
+
 	/**
 	 * If Pop-up Frequency is "Test Mode," assess whether it should be shown.
 	 *
 	 * @param object $popup The popup to assess.
 	 * @return bool Should popup be shown based on Test Mode assessment.
 	 */
 	public static function assess_test_mode( $popup ) {
-		if ( is_user_logged_in() ) {
-			if ( Newspack_Popups::previewed_popup_id() ) {
-				return true;
-			}
-			if ( 'test' !== $popup['options']['frequency'] || ! current_user_can( 'edit_others_pages' ) ) {
-				return false;
-			}
-		} else {
-			if ( 'test' === $popup['options']['frequency'] ) {
-				return false;
-			}
+		if ( 'test' === $popup['options']['frequency'] ) {
+			return is_user_logged_in() && ( current_user_can( 'edit_others_pages' ) || Newspack_Popups::previewed_popup_id() );
 		}
 		return true;
 	}
+
+	/**
+	 * If Pop-up has categories, it should only be shown on posts/pages with those.
+	 *
+	 * @param object $popup The popup to assess.
+	 * @return bool Should popup be shown based on categories it has.
+	 */
+	public static function assess_categories_filter( $popup ) {
```

Comment:
Could you explain the need for this check? I haven't run it down thoroughly, but doesn't the logic in `retrieve_category_overlay_popup` already guarantee that there is category intersection between post and pop-up?

Reply:
`retrieve_category_overlay_popup` does indeed guarantee category intersection between post and pop-up, but just for the _overlay_ popups. This check ensures that an inline popup with a set category will not be shown on a post with a different category.